File: code/anthony/python/solutions/lab1.py
"""
Lab 1: Number to Phrase
Convert a given number into its english representation. For example: 67 becomes 'sixty-seven'. Handle numbers from 0-99.

Hint: you can use modulus to extract the ones and tens digit.

x = 67
tens_digit = x//10
ones_digit = x%10
Hint 2: use the digit as an index for a list of strings OR as a key for a dict of digit:phrase pairs.

Version 2
Handle numbers from 100-999.

Version 3 (optional)
Convert a number to roman numerals.

Version 4 (optional)
Convert a time given in hours and minutes to a phrase.
"""

number_to_phrase = {
    0: "",
    1: "one",
    2: "two",
    3: "three",
    4: "four",
    5: "five",
    6: "six",
    7: "seven",
    8: "eight",
    9: "nine",
    10: "ten",
    11: "eleven",
    12: "twelve",
    13: "thirteen",
    14: "fourteen",
    15: "fifteen",
    16: "sixteen",
    17: "seventeen",
    18: "eighteen",
    19: "nineteen",
    20: "twenty",
    30: "thirty",
    40: "forty",
    50: "fifty",
    60: "sixty",
    70: "seventy",
    80: "eighty",
    90: "ninety",
}

# Get number from user
raw_number = int(input("Enter a number from 0-9999: "))  # 23


if raw_number < 0:
    print("Invalid input.")
elif raw_number == 0:
    print("zero")
elif raw_number <= 20:
    print(number_to_phrase[raw_number])
elif raw_number <= 99:
    tens_digit = raw_number // 10
    tens_digit = tens_digit * 10
    ones_digit = raw_number % 10

    print(number_to_phrase[tens_digit] + " " + number_to_phrase[ones_digit])
    # print(f"{number_to_phrase[tens_digit]} {number_to_phrase[ones_digit]}")
elif raw_number <= 999:

    hundreds_digit = raw_number // 100

    # 123
    tens_digit = raw_number // 10  # 12
    tens_digit = tens_digit % 10  # 2
    tens_digit = tens_digit * 10  # 20

    ones_digit = raw_number % 10

    print(number_to_phrase[hundreds_digit] + " hundred and " +
          number_to_phrase[tens_digit] + " " + number_to_phrase[ones_digit])

elif raw_number <= 9999:
    # 1234
    thousands_digit = raw_number // 1000  # 1
    hundreds_digit = raw_number // 100  # 12
    hundreds_digit = hundreds_digit % 10  # 2
    tens_digit = raw_number // 10  # 123
    tens_digit = tens_digit % 10  # 3
    tens_digit = tens_digit * 10  # 30
    ones_digit = raw_number % 10  # 4

    print(number_to_phrase[thousands_digit] + " thousand " + number_to_phrase[hundreds_digit] + " hundred and " +
          number_to_phrase[tens_digit] + " " + number_to_phrase[ones_digit])

else:
    print("Not yet implemented")

File: code/adam/javascript/solutions/lab1.js


const numbers = 
{0:"",
1:"one", 
2:"two", 
3:"three", 
4:"four",
5:"five", 
6:"six", 
7:"seven", 
8:"eight", 
9:"nine", 
10:"ten",
11:"eleven", 
12:"twelve", 
13:"thirteen", 
14:"fourteen", 
15:"fifteen", 
16:"sixteen",
17:"seventeen",
18:"eighteen",
19:"nineteen",
20:"twenty",
30:"thirty",
40:"fourty",
50:"fifty",
60:"sixty",
70:"seventy",
80:"eighty",
90:"ninety"
}

// allows user to pick their own number to change it to phrase form
let pickNumber = Number(prompt("Enter a number between 1-999: "))

function conversion(pickNumber){
    if(pickNumber in numbers) return numbers[pickNumber];

    let words = '';


    //  hundreds place
    if(pickNumber >= 100){
        words = words + conversion(Math.floor(pickNumber / 100)) + " hundred";
        pickNumber = pickNumber % 100;
    }

    // if number is greater than zero add the "and" to the phrase
    if(pickNumber > 0){
        if(words !== "") words += " and ";
        

        if(pickNumber < 20) words += numbers[pickNumber];
        
        //  tens place
        else{
            words += numbers[Math.floor(pickNumber / 10) * 10];
            
            //  ones place
            if (pickNumber % 10 > 0){
                // concat words + the numbers from dict and add "-" between the tens and ones place
                words += "-" + numbers[pickNumber % 10]
            }
        }
    }

    return words;
}
console.log(conversion(pickNumber))
